src/dmd/lexer.d

```diff
@@ -2192,7 +2214,239 @@ class Lexer
+
+                default:
+                    break Lsuffixes;
+            }
+            ++p;
+            if ((flags & f) && !err)
+            {
+                error("duplicate integer suffixes");
+                err = true;
+            }
+            flags = cast(FLAGS)(flags | f);
+        }
+
+        void overflow()
+        {
+            error("integer overflow");
+        }
+
+        TOK result = TOK.int32Literal;     // default
+        switch (flags)
+        {
+            /* Since D doesn't have a variable sized `long` or `unsigned long` type,
+             * this code deviates from C by picking D int, uint, long, or ulong instead
+             */
+
+            case FLAGS.octalhex:
+                /* Octal or Hexadecimal constant.
+                 * First that fits: int, unsigned, long, unsigned long,
+                 * long long, unsigned long long
+                 */
+                if (longsize == 4)
+                {
+                    if (n & 0x8000000000000000L)
+                        result = TOK.uns64Literal;
+                    else if (n & 0xFFFFFFFF00000000L)
+                        result = TOK.int64Literal;
+                    else if (n & 0x80000000)
+                        result = TOK.uns32Literal;
+                    else
+                        result = TOK.int32Literal;
+                }
+                else
+                {
+                    if (n & 0x8000000000000000L)
+                        result = TOK.uns64Literal;      // unsigned long
+                    else if (n & 0xFFFFFFFF00000000L)
+                        result = TOK.int64Literal;      // long
+                    else if (n & 0x80000000)
+                        result = TOK.uns32Literal;
+                    else
+                        result = TOK.int32Literal;
+                }
+                break;
+
+            case FLAGS.decimal:
+                /* First that fits: int, long, long long
+                 */
+                if (longsize == 4)
+                {
+                    if (n & 0x8000000000000000L)
+                        result = TOK.uns64Literal;
+                    else if (n & 0xFFFFFFFF80000000L)
+                        result = TOK.int64Literal;
+                    else
+                        result = TOK.int32Literal;
+                }
+                else
+                {
+                    if (n & 0x8000000000000000L)
+                        result = TOK.uns64Literal;      // unsigned long
+                    else if (n & 0xFFFFFFFF80000000L)
+                        result = TOK.int64Literal;      // long
+                    else
+                        result = TOK.int32Literal;
+                }
+                break;
+
+            case FLAGS.octalhex | FLAGS.unsigned:
+            case FLAGS.decimal | FLAGS.unsigned:
+                /* First that fits: unsigned, unsigned long, unsigned long long
+                 */
+                if (longsize == 4)
```

Comment:
Both branches are identical.

Reply:
See the comment on the second branch. It's like this if someone wants to fix the obscure problem that C `long` isn't a unique type.